### PR TITLE
Added code for generating forecast and refacored code.

### PR DESCRIPTION
**Pull request by g-bhardwaj**
Please review and merge.

### CODE REVIEW

File: eda/data_analysis.py [hunk 1]
```diff
@@ -104,9 +122,82 @@ def prepare_data(df: pd.DataFrame) -> pd.DataFrame:
...
+
+    Returns:
+        Generate the correct date range to be used to train models.
+
+    """
+    category_sector_spend = df[(df["Category"] == category) & (df["MarketSector"] == sector)].reset_index(drop=True)
+    category_sector_spend.set_index("SpendMonth", inplace=True, drop=False)
+    # complete_data = pd.DataFrame(data=pd.date_range(start=category_sector_spend["SpendMonth"].min(axis=0),
+    #                                                 end=category_sector_spend["SpendMonth"].max(axis=0), freq="MS"),
+    #                              columns=["SpendMonth"])
```
Comment (kapppa-joe) [inline eda/data_analysis.py]: can we remove this commented out code if it is not needed?

File: eda/data_analysis.py [hunk 2]
```diff
@@ -104,9 +122,82 @@ def prepare_data(df: pd.DataFrame) -> pd.DataFrame:
...
+               )
+               < DATEADD(month,-2,GETDATE())
+               GROUP BY spend.Category,
+               cust.MarketSector,
+               spend.FYMonthKey
+            )
+            AS a
+            GROUP BY a.Category,a.MarketSector
+            ORDER BY RecordCount DESC,a.Category,a.MarketSector
+            """
```
Comment (kapppa-joe) [inline eda/data_analysis.py]: Here we sort the combinations by record count. When I run this SQL query, the outcome seems to have combinations of MarketSector = Unassigned at the top. Is this expected? 

when I run this sql separately, the result looks like this:

"RecordCount","Category","MarketSector"
"205","Network Services","Unassigned"
"163","Financial Services","Unassigned"
"150","Document Management & Logistics","Unassigned"
"147","Technology Products & Services","Unassigned"
"130","Fleet","Unassigned"
"122","Workforce Health & Education","Unassigned"
"116","Construction","Defence and Security"
"116","Construction","Education"
"116","Construction","Government Policy"

File: eda/data_analysis.py [hunk 3]
```diff
@@ -123,164 +214,13 @@ def get_category_sector(index: int = 0) -> Tuple[str, str]:
...
-        # ("Managed Service", "Education"),
-        # ("Below Threshold", "Defence and Security"),
-        # ("Below Threshold", "Local Community and Housing"),
-        # ("Below Threshold", "Government Policy"),
-        # ("Below Threshold", "Education"),
-        # ("Below Threshold", "Culture, Media and Sport")
-    ]
+
+    if not os.path.exists(local_category_sector_path):
+        get_all_category_sector(latest=True)
```
Comment (kapppa-joe) [inline eda/data_analysis.py:244]: When the local file doesn't exist and get into line 219, then probably the user can't easily predict which combinations they will get by the index number they input. 
Would it be good if we change this to save the result of get_all_category_sector to local file, then ask the user re-run the script after checking the index number with local file?

File: eda/data_analysis.py (hunk 1 above)
Comment (g-bhardwaj) [reply eda/data_analysis.py]: Good catch. Fixed!

File: eda/data_analysis.py (hunk 2 above)
Comment (g-bhardwaj) [reply eda/data_analysis.py]: Good point, though it is not going to matter on the count of Category and MarketSector, but I modified the query.
Fixed!

File: eda/data_analysis.py (hunk 3 above)
Comment (g-bhardwaj) [reply eda/data_analysis.py:244]: Generally no one remembers all the categories and sectors, so this method is just a helper method that user can use find a combination of category. It doesn't matter much from where it is coming. 99% of the time when data scientist wants to fine tune anything he won't be relying on this method instead he will be using something like this:
category, sector = ("Construction","Education") i.e. passing hard core values to see how particular combination is behaving. So this doesn't require change.

**Pull request merged**